pkg/3scale/amp/operator/backend_options_provider.go

```diff
@@ -0,0 +1,144 @@
+		backendOptions: component.NewBackendOptions(),
+		secretSource:   NewSecretSource(client, namespace),
+	}
+}
+
+func (o *OperatorBackendOptionsProvider) GetBackendOptions() (*component.BackendOptions, error) {
+	o.backendOptions.AppLabel = *o.apimanagerSpec.AppLabel
+	o.backendOptions.TenantName = *o.apimanagerSpec.TenantName
+	o.backendOptions.WildcardDomain = o.apimanagerSpec.WildcardDomain
+
+	err := o.setSecretBasedOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	o.setResourceRequirementsOptions()
+	o.setReplicas()
+
+	err = o.backendOptions.Validate()
+	return o.backendOptions, err
+}
+
+func (o *OperatorBackendOptionsProvider) setSecretBasedOptions() error {
+	cases := []struct {
+		field       *string
+		secretName  string
+		secretField string
+		defValue    string
+	}{
+		{
+			&o.backendOptions.SystemBackendUsername,
+			component.BackendSecretInternalApiSecretName,
+			component.BackendSecretInternalApiUsernameFieldName,
+			component.DefaultSystemBackendUsername(),
+		},
+		{
+			&o.backendOptions.SystemBackendPassword,
+			component.BackendSecretInternalApiSecretName,
+			component.BackendSecretInternalApiPasswordFieldName,
+			component.DefaultSystemBackendPassword(),
+		},
+		{
+			&o.backendOptions.ServiceEndpoint,
+			component.BackendSecretBackendListenerSecretName,
+			component.BackendSecretBackendListenerServiceEndpointFieldName,
+			component.DefaultBackendServiceEndpoint(),
+		},
+		{
+			&o.backendOptions.RouteEndpoint,
+			component.BackendSecretBackendListenerSecretName,
+			component.BackendSecretBackendListenerRouteEndpointFieldName,
+			fmt.Sprintf("https://backend-%s.%s", *o.apimanagerSpec.TenantName, o.apimanagerSpec.WildcardDomain),
+		},
+		{
+			&o.backendOptions.StorageURL,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisStorageURLFieldName,
+			component.DefaultBackendRedisStorageURL(),
+		},
+		{
+			&o.backendOptions.QueuesURL,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisQueuesURLFieldName,
+			component.DefaultBackendRedisQueuesURL(),
+		},
+		{
+			&o.backendOptions.StorageSentinelHosts,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisStorageSentinelHostsFieldName,
+			"",
+		},
+		{
+			&o.backendOptions.StorageSentinelRole,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisStorageSentinelRoleFieldName,
+			"",
+		},
+		{
+			&o.backendOptions.QueuesSentinelHosts,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisQueuesSentinelHostsFieldName,
+			"",
+		},
+		{
+			&o.backendOptions.QueuesSentinelRole,
+			component.BackendSecretBackendRedisSecretName,
+			component.BackendSecretBackendRedisQueuesSentinelRoleFieldName,
+			"",
+		},
+	}
+
+	for _, option := range cases {
+		val, err := o.secretSource.FieldValue(option.secretName, option.secretField, option.defValue)
```

Comment:
In setSecretBasedOptions, for a secret that has N fields that secret is "getted" N times.
Before it was only one time. It might be good enough as I think operator-sdk does some sort of caching but if the operator-sdk that the caching has does not work as we expect this means that it makes lot more API calls than before.

Reply:
`SecretSource` struct should implement caching. There is a `// TODO implement caching?? ` comment there to start discussion. I think is should implement caching. Very simple implementation as is it assumed that a new `SecretSource` object will be instantiated for each option provider and for each reconciliation loop. So no need to implement size limit, expiration or any eviction action.

Reply:
I think we should implement it too or change it to get the secret at the beginning for all fields.

Leaving it as it is right now I think it might be dangerous. Keep in mind that right now for each get we might receive different resourceVersions of the object so in the same method we could receive "inconsistent" states of the fields among them

Reply:
Ok, I will implement a simple caching solutions. Caching life-cycle will be limited to the reconciliation loop and each options provider will have its own caching instance.

Reply:
:+1: I think it makes sense at this moment